internal/broker/data.go

```diff
@@ -0,0 +1,58 @@
+package broker
+
+import (
+	"bytes"
+	"encoding/json"
+	"text/template"
+
+	"github.com/ubuntu/oidc-broker/internal/providers/group"
+)
+
+// userInfo represents the user information that is returned to authd.
+type userInfo struct {
+	Name   string
+	UUID   string
+	Home   string
+	Shell  string
+	Gecos  string
+	Groups []group.Info
+}
+
+// MarshalJSON implements the json.Marshaler interface for the userInfo type.
+func (u userInfo) MarshalJSON() ([]byte, error) {
+	var buf bytes.Buffer
+	err := template.Must(template.New("").Parse(`{"userinfo": {
+		"name": "{{.Name}}",
+		"uuid": "{{.UUID}}",
+		"gecos": "{{.Gecos}}",
+		"dir": "{{.Home}}",
+		"shell": "{{.Shell}}",
+		"groups": [ {{range $index, $g := .Groups}}
+			{{- if $index}}, {{end -}}
+			{"name": "{{.Name}}", "ugid": "{{.UGID}}"}
+		{{- end}} ]
+}}`)).Execute(&buf, u)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
```

Comment:
To understand, what is the rationale for having this (which is is still technically still potentially affected by the root issue), compared to native marshalling?

Reply:
Honestly, I don't like it either. There was a problem with it, but I think it used to happen before we defined a type for the user groups. I'm down to strip it and run some checks to see if everything still works as expected.